openprescribing/frontend/price_per_unit/substitution_sets.py

```diff
@@ -0,0 +1,274 @@
+
+        ([<list of generic BNF codes], [<list of formulations>])
+
+    For instance, Tramadol tablets and capsules are distinct generic
+    presentations, but are almost always substitutable for each other so the
+    response might include:
+
+        # 100mg tablets and capsules
+        (['040702040AAACAC', '040702040AAAHAH'], ['Tab', 'Cap'])
+
+        # 200mg tablets and capsules
+        (['040702040AAAEAE', '040702040AAAJAJ'], ['Tab', 'Cap'])
+
+    The swaps file must be manually curated to ensure that the substitutions
+    are clinically sensible.
+    """
+    with open(filename, "rt") as handle:
+        data = list(csv.DictReader(handle))
+    code_pairs = []
+    formulations = {}
+    for row in data:
+        if row["Really equivalent?"].strip() != "Y":
+            continue
+        code = row["Code"].strip()
+        alternative_code = row["Alternative code"].strip()
+        # Check that all codes in the substitutions table are generics
+        assert code == generic_equivalent_for_bnf_code(code)
+        assert alternative_code == generic_equivalent_for_bnf_code(alternative_code)
+        assert code != alternative_code
+        code_pairs.append((code, alternative_code))
+        formulations[code] = row["Formulation"].strip()
+        formulations[alternative_code] = row["Alternative formulation"].strip()
+    # We combine pairs of equivalent codes into groups of equivalent codes.
+    # This means if our substitutions table says "A and B are equivalent" and
+    # "B and C are equivalent" then we can handle this correctly.
+    for code_group in groups_from_pairs(code_pairs):
+        # Get the unique formulations involved in this group
+        formulations_for_group = [
+            f for f in set(map(formulations.get, code_group)) if f
+        ]
+        yield code_group, formulations_for_group
+
+
+def groups_from_pairs(pairs):
+    """
+    Accepts a list of pairs and combines any overlapping pairs into groups
+
+    >>> list(groups_from_pairs([
+    ...     (1, 2),
+    ...     (3, 4),
+    ...     (5, 6),
+    ...     (1, 3),
+    ... ]))
+    [[1, 2, 3, 4], [5, 6]]
+
+    In set theoretic terms, `pairs` is an equivalence relation and `groups` are
+    the equivalence classes it induces.
+    """
+    groups = {}
+    for element, other_element in pairs:
+        group = groups.setdefault(element, [element])
+        other_group = groups.get(other_element, [other_element])
+        if other_group is not group:
+            group.extend(other_group)
+            for member in other_group:
+                groups[member] = group
+    for element, group in groups.items():
+        # Each group will occur multiple times in this loop, once for each of
+        # its elements. But we want to return just the unique groups, so we
+        # pick (arbitrarily) the first element of each group to identify it
+        # with.
+        if element == group[0]:
+            yield group
```

Comment:
This doesn't appear to be independently tested -- it's a little hairy, so could it be?